cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/k8s/UserServiceProgramMain.java

```diff
@@ -0,0 +1,409 @@
+    private final CompletableFuture<ProgramController> controllerFuture;
+    private final CompletableFuture<ProgramController.State> programCompletion;
+    private final CConfiguration cConf;
+    private final ProgramRunner programRunner;
+    private final ArtifactFinder artifactFinder;
+    private final ApplicationSpecification appSpec;
+    private final ProgramOptions programOptions;
+    private final ProgramRunId programRunId;
+    private final String bindHost;
+    private final long maxStopSeconds;
+    private File pluginsDir;
+    private Program program;
+
+    private UserService(CConfiguration cConf, ArtifactFinder artifactFinder, ProgramRunner programRunner,
+                        ApplicationSpecification appSpec, ProgramOptions programOptions, String bindHost) {
+      this.controllerFuture = new CompletableFuture<>();
+      this.programCompletion = new CompletableFuture<>();
+      this.cConf = cConf;
+      this.programRunner = programRunner;
+      this.artifactFinder = artifactFinder;
+      this.appSpec = appSpec;
+      this.programOptions = programOptions;
+      this.bindHost = bindHost;
+      this.programRunId = programOptions.getProgramId().run(ProgramRunners.getRunId(programOptions));
+      this.maxStopSeconds = cConf.getLong(Constants.AppFabric.PROGRAM_MAX_STOP_SECONDS, 60);
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+      // fetch the app artifact
+      NamespaceId programNamespace = programRunId.getNamespaceId();
+      Set<ArtifactId> pluginArtifacts = new HashSet<>();
+      for (Plugin plugin : appSpec.getPlugins().values()) {
+        pluginArtifacts.add(convert(programNamespace, plugin.getArtifactId()));
+      }
+
+      // need to set the plugins directory in system args, as well as the host to bind to
+      File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR));
+      pluginsDir = new File(tmpDir, "plugins");
+      Map<String, String> systemArgs = new HashMap<>(programOptions.getArguments().asMap());
+      systemArgs.put(ProgramOptionConstants.PLUGIN_DIR, pluginsDir.getAbsolutePath());
+      systemArgs.put(ProgramOptionConstants.HOST, bindHost);
+      // TODO: (CDAP-15018) figure out scaling
+      systemArgs.put(ProgramOptionConstants.INSTANCE_ID, "1");
+      systemArgs.put(ProgramOptionConstants.INSTANCES, "1");
+      ProgramOptions updatedOptions = new SimpleProgramOptions(programRunId.getParent(), new BasicArguments(systemArgs),
+                                                               programOptions.getUserArguments());
+
+      File programDir = new File(tmpDir, "program");
+      DirUtils.mkdirs(programDir);
+      File programJarFile = new File(programDir, "program.jar");
+      Location programJarLocation =
+        artifactFinder.getArtifactLocation(convert(programNamespace, appSpec.getArtifactId()));
+      Files.copy(programJarLocation.getInputStream(), programJarFile.toPath());
```

Comment:
This will leave an unclosed InputStream. Use `Locations.linkOrCopy` instead.

Reply:
done